CSC8503CoreClasses/GuardObject.cpp

```diff
@@ -2,42 +2,175 @@
 #include "PlayerObject.h"
 #include "Ray.h"
 #include "Debug.h"
+#include "PhysicsObject.h"
+#include "BehaviourSelector.h"
 
 using namespace NCL;
 using namespace CSC8503;
 
 GuardObject::GuardObject(const std::string& objectName) {
 	mName = objectName;
+	mRootSequence = new BehaviourSequence("Root Sequence");
+	CanSeePlayer = false;
+	HasCaughtPlayer = false;
+	HasConfiscatedItems = false;
+	BehaviourTree();
 }
 
 GuardObject::~GuardObject() {
 	delete mPlayer;
 	delete mWorld;
+	delete mSightedObject;
+	delete mRootSequence;
 }
 
 void GuardObject::UpdateObject(float dt) {
 	RaycastToPlayer();
+	ExecuteBT();
 }
 
 void GuardObject::RaycastToPlayer() {
 	mPlayer = dynamic_cast<PlayerObject*>(mPlayer);
 	Vector3 dir = (mPlayer->GetTransform().GetPosition() - this->GetTransform().GetPosition()).Normalised();
-	RayCollision closestCollision;
-	Ray r = Ray(this->GetTransform().GetPosition(), dir);
 	float ang = Vector3::Dot(dir, AngleOfSight());
 	if (ang > 2) {
+		RayCollision closestCollision;
+		Ray r = Ray(this->GetTransform().GetPosition(), dir);
 		if (mWorld->Raycast(r, closestCollision, true, this)) {
```

Comment:
Is there a way to reduce if statements here?
What if we combined the one in ln 36 and 39 to 
if (ang>2 && mWorld->Raycast(r, closestCollision, true, this))?

Reply:
I split it into two so the raycast didn't run if the player wasn't in front of the guard. So it wont be raycasting constantly even when the player isn't seen.